File: 93.复原ip地址.py
#
# @lc app=leetcode.cn id=93 lang=python3
#
# [93] 复原IP地址
#

# @lc code=start
from typing import List


class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = []
        length = len(s)

        def back(li: List, start, end, depth):
            if start > length - 1:
                return
            # li = li + []
            if depth == 3:
                if int(s[start:]) <= 255:
                    if start < length - 1 and s[start] == '0':
                        return
                    li.append(s[start:])
                    res.append('.'.join(li))
                    li.pop()
                return
            else:
                if end > length or int(s[start:end]) > 255:
                    return
                li.append(s[start:end])
                back(li, end, end+1, depth+1)
                li.pop()
                if s[start] != '0':
                    back(li, start, end + 1, depth)
        back([], 0, 1, 0)
        return res

# @lc code=end

File: 93.复原ip地址.java
/*
 * @lc app=leetcode.cn id=93 lang=java
 *
 * [93] 复原IP地址
 */

// @lc code=start
import java.util.ArrayList;
class Solution {
    ArrayList<String> res = new ArrayList<>();
    ArrayList<String> li = new ArrayList<>();
    int len = 0;
    String s = "";
    public List<String> restoreIpAddresses(String s) {
        this.len = s.length();
        this.s = s;
        backtrack(0, 1, 0);
        return res;
    }
    private void backtrack(int start, int end, int depth) {
        if (start > len-1)
            return;
        if (depth == 3) {
            if (len - start > 3)
                return;
            int tmp = Integer.parseInt(s.substring(start));
            if (start < len -1 && s.charAt(start) == '0')
                return;
            if (tmp <= 255) {
                li.add(s.substring(start));
                res.add(geneString());
                li.remove(depth);
            }
            return;
        } else {
            if (end > len - 1 || Integer.parseInt(s.substring(start, end)) > 255)
                return;
            li.add(s.substring(start, end));
            backtrack(end, end + 1, depth + 1);
            li.remove(depth);
            if (s.charAt(start) != '0')
                backtrack(start, end+1, depth);
        }
    }

    private String geneString() {
        StringBuilder builder = new StringBuilder(li.get(0));
        for (int i = 1; i < li.size(); i++) {
            builder.append('.');
            builder.append(li.get(i));
        }
        return builder.toString();
    }

}
// @lc code=end

